### PR TITLE
Avoid watchlist to crash with special cases

### PR DESCRIPTION
**Pull request by saraycp**
Right now, the new watchlist implementation fails when `@project` or `@package` don't contain an object but a string or nil. 
This happens when we have links and multibuild packages.

With the introduced code, we make sure we always get an object when this is possible. Otherwise, we assume they are nil, so we don't offer an item to add to the watchlist. We decided to add that logic inside the `WatchlistComponent` because it would be too much ruby code for the layout view.

In the case of inherited packages belonging to project links, we decided not to provide the link to add the package to the watchlist. It's not straightforward to keep track of the linked project.

After this is merged, we can enable back the `new_watchlist` feature flag in production. First for the `staff` group and then for the `beta` group.

### CODE REVIEW

File: src/api/app/views/layouts/webui/webui.html.haml [hunk 1]
```diff
@@ -77,9 +77,22 @@
...
       - if User.session
         - if feature_enabled?(:new_watchlist)
+          :ruby
+            # NOTE: temporary fix to avoid watchlist crash with special packages or projects (remote, links, multibuild)
+            project = @project if @project.is_a?(Project)
+            package = @package if project && @package.is_a?(Package)
+
+            # NOTE: For a while, we'll track in the log those special cases for future improvements.
+            if (@project && !project) || (@package && !package)
+              Rails.logger.debug do
```
Comment (hennevogel) [inline src/api/app/views/layouts/webui/webui.html.haml:87]: We don't do debug logs in production

Comment (saraycp) [reply src/api/app/views/layouts/webui/webui.html.haml]: This was removed with the last changes.

**Comment by dmarcoux** [comment]
I'm against this change, because as [I wrote in a previous PR which was quite similar](https://github.com/openSUSE/open-build-service/pull/12245#discussion_r812658017):

> If we provide a package which isn't really a package, but a string to this component, we should change the code where this component is instantiated.
>
>A view component should deal with view code, not fetching data from the database. (...)

What is preventing us from repeating the same approach as #12245? This would imply refactoring the controllers which are returning strings instead of a package/project/bs_request.

**Review by dmarcoux — DISMISSED** [review]
I'm against this change and I refer to a comment [I wrote in a previous PR which was quite similar](https://github.com/openSUSE/open-build-service/pull/12245#discussion_r812658017):

> If we provide a package which isn't really a package, but a string to this component, we should change the code where this component is instantiated.
>
>A view component should deal with view code, not fetching data from the database. (...)

What is preventing us from repeating the same approach as #12245? This would imply refactoring the controllers which are returning strings instead of a package/project/bs_request. I agree that the layout view shouldn't handle this logic, this is why it belongs in the controllers, not in view code.

File: src/api/app/components/watchlist_component.rb [hunk 2]
```diff
@@ -11,16 +11,36 @@ class WatchlistComponent < ApplicationComponent
     'BsRequest' => 'Watch this request'
   }.freeze
 
-  def initialize(user:, project: nil, package: nil, bs_request: nil)
+  def initialize(user:, bs_request: nil, package: nil, project: nil)
```
Comment (dmarcoux) [inline src/api/app/components/watchlist_component.rb:14]: What is the reason behind this change?

Comment (eduardoj) [reply src/api/app/components/watchlist_component.rb:14]: To arrange the parameters in the same order they will be called inside the initialize, a couple of lines below:
```
@object_to_be_watched = object_to_be_watched(bs_request, package, project)
```

Comment (dmarcoux) [reply src/api/app/components/watchlist_component.rb:14]: Humm... alright.

**Comment by dmarcoux** [comment]
It doesn't feel good to patch the view component, it's definitely adding technical debt instead of solving the core issue in the controllers. How many controllers are causing issues?

**Comment by eduardoj** [comment]
> It doesn't feel good to patch the view component, it's definitely adding technical debt instead of solving the core issue in the controllers. How many controllers are causing issues?

That's the question. If we patch the view component without adding any alert or log, as it is currently proposed by the PR, I feel we lose the chance of knowing what to fix.

If we don't do debug logs in production, could we just change that line to this one?
```
Rails.logger.info do
```
Would this be fine for you, @hennevogel?

**Comment by hennevogel** [comment]
> It doesn't feel good to patch the view component, it's definitely adding technical debt instead of solving the core issue in the controllers. How many controllers are causing issues?

There is not core issue in controllers. There is an architecture (Links) that allows this to happen. You will not solve this problem in *any* code on top.

*Please* work with this architecture in the *new* component that you add instead of moving this deeper into the stack.
